RecoLocalCalo/EcalRecProducers/plugins/EcalUncalibRecHitPhase2WeightsKernels.cu

```diff
@@ -0,0 +1,62 @@
+#include <cuda.h>
+
+#include "FWCore/Utilities/interface/CMSUnrollLoop.h"
+#include "DataFormats/EcalRecHit/interface/EcalUncalibratedRecHit.h"
+#include "DataFormats/EcalDigi/interface/EcalLiteDTUSample.h"
+#include "DataFormats/EcalDigi/interface/EcalConstants.h"
+
+#include "EcalUncalibRecHitPhase2WeightsKernels.h"
+
+namespace ecal {
+  namespace weights {
+
+    __global__ void Phase2WeightsKernel(uint16_t const* digis_in,
+                                        uint32_t const* dids,
+                                        ::ecal::reco::StorageScalarType* amplitude,
+                                        ::ecal::reco::StorageScalarType* amplitudeError,
+                                        uint32_t* dids_out,
+                                        int const nchannels,
+                                        double* weights,
+                                        uint32_t* flags) {
```

Comment:
Could you try if adding `__restrict__` to the pointer arguments (and `const` if it's missing anywhere) improves the performance ?
If you are lucky, it could even make it unnecessary to copy the data to the shared memory, because the GPU will use it's local cache instead.
```suggestion
    __global__ void Phase2WeightsKernel(uint16_t const* digis_in,
                                        uint32_t const* __restrict__ dids,
                                        ::ecal::reco::StorageScalarType* __restrict__ amplitude,
                                        ::ecal::reco::StorageScalarType* __restrict__ amplitudeError,
                                        uint32_t* __restrict__ dids_out,
                                        int const nchannels,
                                        double const* __restrict__ weights,
                                        uint32_t* __restrict__ flags) {
```

Reply:
Some background information: the compiler does not have any way of knowing that (for example) `dids` and `flags` do not point to overlapping data, so the writes to one might affect the reads form the other.
`__restrict__` tells the compiler that this is _not_ the case, and that the "restricted" pointer is the only way to access the pointed-to data, so the compiler can optimise the code more aggressively.
On recent GPUs, there is an additional benefit that the `const* __restrict__` data _may_ be cached in a "read-only cache" local to the block, which should have similar performance as using the shared memory for explicit caching.